src/fs/bootstrap_record.cc

```diff
@@ -0,0 +1,182 @@
+                    invalid_bootstrap_record("Cluster size should be a power of 2 and be divisible by sector size"));
+        }
+        if (bootstrap_record_disk.shards_nb > max_shards_nb || bootstrap_record_disk.shards_nb == 0) {
+            return make_exception_future<bootstrap_record>(
+                    invalid_bootstrap_record("Shards number is invalid"));
+        }
+
+        const std::vector<shard_info> tmp_shards_info(bootstrap_record_disk.shards_info,
+                bootstrap_record_disk.shards_info + bootstrap_record_disk.shards_nb);
+
+        if (!is_shards_info_valid(tmp_shards_info)) {
+            return make_exception_future<bootstrap_record>(invalid_bootstrap_record("Shards info is invalid"));
+        }
+
+        bootstrap_record bootstrap_record_mem(bootstrap_record_disk.version,
+                bootstrap_record_disk.sector_size,
+                bootstrap_record_disk.cluster_size,
+                bootstrap_record_disk.root_directory,
+                std::move(tmp_shards_info));
+
+        return make_ready_future<bootstrap_record>(std::move(bootstrap_record_mem));
+    });
+}
+
+future<> bootstrap_record::write_to_disk(block_device& device) const {
+    // initial check
+    if (!is_sector_size_valid(sector_size)) {
+        return make_exception_future<>(
+                invalid_bootstrap_record("Sector size should be a power of 2 and greater than 512"));
+    }
+    if (!is_cluster_size_valid(cluster_size, sector_size)) {
+        return make_exception_future<>(
+                invalid_bootstrap_record("Cluster size should be a power of 2 and be divisible by sector size"));
+    }
+    if (shards_nb() > max_shards_nb || shards_nb() == 0) {
+        return make_exception_future<>(
+                invalid_bootstrap_record("Shards number is invalid"));
+    }
+    if (!is_shards_info_valid(shards_info)) {
+        return make_exception_future<>(invalid_bootstrap_record("Shards info is invalid"));
+    }
+
+    bootstrap_record_disk bootstrap_record_disk;
+    // using memset to zero out the whole memory used by bootstrap_record_disk including paddings
+    std::memset(&bootstrap_record_disk, 0, sizeof(bootstrap_record_disk));
+
+    // prepare bootstrap_record_disk records
+    bootstrap_record_disk.magic = bootstrap_record::magic_number;
+    bootstrap_record_disk.version = version;
+    bootstrap_record_disk.sector_size = sector_size;
+    bootstrap_record_disk.cluster_size = cluster_size;
+    bootstrap_record_disk.root_directory = root_directory;
+    bootstrap_record_disk.shards_nb = shards_nb();
+    std::copy(shards_info.begin(), shards_info.end(), bootstrap_record_disk.shards_info);
+    bootstrap_record_disk.crc = crc32(&bootstrap_record_disk, crc_offset);
+
+    auto bootstrap_record_buff = temporary_buffer<char>::aligned(alignment, aligned_bootstrap_record_size);
+    std::memcpy(bootstrap_record_buff.get_write(), &bootstrap_record_disk, sizeof(bootstrap_record_disk));
```

Comment:
Why prepare everything on stack and then copy it to a heap-allocated buffer instead of preparing everything in the buffer from the start?
```cpp
    auto bootstrap_record_buff = temporary_buffer<char>::aligned(alignment, aligned_bootstrap_record_size);
    bootstrap_record_disk* bootstrap_record_disk = (bootstrap_record_disk*)bootstrap_record_buff.get_write();
    bootstrap_record_disk->magic = (...)
    (...)
```
The buffer is properly aligned, so this unsafe casting is actually quite safe. It will also remove the need of one of the memsets.